### PR TITLE
Fix #7671: add `PreserveWhitespace` flag to `XslTransformation`

### LINKED ISSUE #7671: Task XslTransformation: How to specify XmlSpace.Preserve?
### Issue Description
The MSBuild Task XslTransformation ([docs](https://docs.microsoft.com/en-us/visualstudio/msbuild/xsltransformation-task?view=vs-2022), [source](https://github.com/dotnet/msbuild/blob/main/src/Tasks/XslTransformation.cs)) always loads the provided XSLT file with `XmlSpace.Default`. I would like to use `XmlSpace.Preserve` instead.

### Steps to Reproduce
See [this project](https://github.com/NMertsch/msbuild-xsltransformation-issue). It contains an XSL stylesheet, an input XML file, and a dotnet project that generates two output files: one with the XslTransformation task, one in `Program.cs`. It also contains both output files so you can see the difference in GitHub.

You can clone the repo, delete the generated files, and run `dotnet run` to re-generate them.

### Expected Behavior
Both files have the same content.

### Actual Behavior
The MSBuild Task does not preserve whitespace and offers no way to change this behavior (at least none that I was able to find).

### Analysis
The desired behavior can be achieved by changing `new XPathDocument(reader)` to `new XPathDocument(reader, XmlSpace.Preserve)` in [this line](https://github.com/dotnet/msbuild/blob/bfd80ab1692721767c3fa8c0d4bcd35254489086/src/Tasks/XslTransformation.cs#L446).

I'm happy that `XmlSpace.Default` is the default. But I'd like to have an option to use `XmlSpace.Preserve` instead, e.g. via `<XslTransformation PreserveWhitespace="true" ... />`.

### Versions & Configurations
To reproduce this, I use the dotnet CLI without IDE.
- `dotnet --version`: 6.0.300
- OS: Windows 10 Pro (Version 10.0.19044 Build 19044)

### PR DESCRIPTION
**Pull request by shadyalfred**
`PreserveWhitespace` option to use `XmlSpace.Preserve` with
`XPathDocument`.

Fixes #7671

### Context
Should fix whitespace issue with XSLT files

### Changes Made
Adding a new public property to the class `XslTransformation`. If it `PreserveWhitespace` is set to true, the produced file will preserve whitespaces.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7716) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7716) before we can accept your contribution.<br/><br/>:x: shadyalfred [sign now](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7716)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/msbuild?pullRequest=7716) it.</sub>

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7716) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7716) before we can accept your contribution.<br/><br/>:x: shadyalfred [sign now](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7716)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/msbuild?pullRequest=7716) it.</sub>

File: src/Tasks/PublicAPI/net/PublicAPI.Shipped.txt [hunk 1]
```diff
@@ -2242,6 +2242,8 @@ Microsoft.Build.Tasks.XslTransformation.XslCompiledDllPath.get -> Microsoft.Buil
 Microsoft.Build.Tasks.XslTransformation.XslCompiledDllPath.set -> void
 Microsoft.Build.Tasks.XslTransformation.XslContent.get -> string
 Microsoft.Build.Tasks.XslTransformation.XslContent.set -> void
+Microsoft.Build.Tasks.XslTransformation.PreserveWhitespace.get -> bool
+Microsoft.Build.Tasks.XslTransformation.PreserveWhitespace.set -> void
```
Comment (rainersigwald) [inline src/Tasks/PublicAPI/net/PublicAPI.Shipped.txt:2246]: These new lines should go into `PublicAPI.Unshipped.txt`, since they haven't (yet!) gone into a public release. The distinction helps us understand what public API we can modify (things that haven't yet released), and what we can't (changing or removing things that have been released would be a breaking change).

If you use Visual Studio or another IDE that can handle Roslyn code fixers, this should be available as a "quick fix" action.

**Review by rainersigwald — COMMENTED** [review]
This change looks great. Can you please add some tests in `src\Tasks.UnitTests\XslTransformation_Tests.cs`?

File: src/Tasks/PublicAPI/net/PublicAPI.Shipped.txt (hunk 1 above)
Comment (shadyalfred) [reply src/Tasks/PublicAPI/net/PublicAPI.Shipped.txt]: I'm using vscode, so I edited them manually. I moved them from `PublicAPI.Shipped.txt` to `PublicAPI.Unshipped.txt`

**Comment by shadyalfred** [comment]
I'm not sure how to test it. I've tried this 982bf28b34e0ffdb3ed4379106f2503f88f8c465 but I don't think it's the correct way to test it

File: src/Tasks.UnitTests/XslTransformation_Tests.cs [hunk 2]
```diff
@@ -848,6 +863,57 @@ public void XsltDocumentThrowsError()
...
+                t.Parameters = _xslParameters;
+                t.PreserveWhitespace = true;
+
+                t.Execute();
+                Console.WriteLine(engine.Log);
+                using (StreamReader sr = new StreamReader(t.OutputPaths[0].ItemSpec))
+                {
+                    string fileContents = sr.ReadToEnd();
+                    Assert.Equal(
+@"<?xml version=""1.0"" encoding=""utf-8""?>
```
Comment (Forgind) [inline src/Tasks.UnitTests/XslTransformation_Tests.cs:896]: To make this a more robust test, maybe make weird whitespace changes in the input and output files?

Comment (shadyalfred) [reply src/Tasks.UnitTests/XslTransformation_Tests.cs:896]: I've tried to add another xml document with extra whitespaces
```csharp
        private readonly string _xmlDocument3 =
@"<root
    Name=""param1""
    Value=""value111"">

    <abc>

        <cde/>


    </abc>

</root>";
```
But toggling `PreserveWhitespace` from true to false doesn't change the output.
It's always the same
```xml
<?xml version="1.0" encoding="utf-8"?>
<surround>
  <root>
    <surround Name="param1" />
    <surround Value="value111" />
    <surround>

    </surround>
    <surround>
      <abc>
        <surround>

        </surround>
        <surround>
          <cde />
        </surround>
        <surround>


    </surround>
      </abc>
    </surround>
    <surround>

</surround>
  </root>
</surround>
```

Maybe only certain xsl transformations mess up the whitespaces. I've tried to look up the [xsl file](https://github.com/NMertsch/msbuild-xsltransformation-issue/blob/main/xsl/fdl2proto.xsl) used in the project where this issue happens but it was too complicated to add as a test case, and I think it imports some libraries or something.

Comment (shadyalfred) [reply src/Tasks.UnitTests/XslTransformation_Tests.cs:896]: I think the surround tags are just what the xsl document, which is provided in the Unit Test file, does.
```csharp
private readonly string _xslDocument = "<xsl:stylesheet version=\"1.0\" xmlns:xsl=\"http://www.w3.org/1999/XSL/Transform\" xmlns:msxsl=\"urn:schemas-microsoft-com:xslt\" exclude-result-prefixes=\"msxsl\"><xsl:output method=\"xml\" indent=\"yes\"/><xsl:template match=\"@* | node()\"><surround><xsl:copy><xsl:apply-templates select=\"@* | node()\"/></xsl:copy></surround></xsl:template></xsl:stylesheet>";
 ```
 I haven't worked with xsl files before, so I'm not sure. Can you link some docs about them so that I can understand the [xsl file](https://github.com/NMertsch/msbuild-xsltransformation-issue/blob/main/xsl/fdl2proto.xsl) better? I think the transformation mentioned in the [issue](https://github.com/dotnet/msbuild/issues/7671) is about `fdl2proto`, not sure what are these tho.

Comment (shadyalfred) [reply src/Tasks.UnitTests/XslTransformation_Tests.cs]: I've added the documents used [here](https://github.com/NMertsch/msbuild-xsltransformation-issue) and used it in the test. Now toggling the `PreserveWhitespace` makes the test fail or pass.

**Comment by Forgind** [comment]
Thanks @shadyalfred!
